File: julia/src/models/columnar.jl
"""Columnar model — analytical aggregations via COLUMNAR_* SQL functions."""

struct ColumnarModel
    conn::LibPQ.Connection
    features::NucleusFeatures
end

"""COLUMNAR_INSERT(table, values_json) → nothing"""
function insert!(m::ColumnarModel, table::String, values::Dict)
    require_nucleus(m.features, "Columnar")
    json_str = JSON3.write(values)
    LibPQ.execute(m.conn, "SELECT COLUMNAR_INSERT(\$1, \$2)", [table, json_str])
    return nothing
end

"""COLUMNAR_COUNT(table) → Int64"""
function columnar_count(m::ColumnarModel, table::String)::Int64
    require_nucleus(m.features, "Columnar")
    result = LibPQ.execute(m.conn, "SELECT COLUMNAR_COUNT(\$1)", [table])
    return _int(result)
end

"""COLUMNAR_SUM(table, column) → Float64"""
function columnar_sum(m::ColumnarModel, table::String, column::String)::Float64
    require_nucleus(m.features, "Columnar")
    result = LibPQ.execute(m.conn, "SELECT COLUMNAR_SUM(\$1, \$2)", [table, column])
    v = _float(result)
    return v === nothing ? 0.0 : v
end

"""COLUMNAR_AVG(table, column) → Float64"""
function columnar_avg(m::ColumnarModel, table::String, column::String)::Float64
    require_nucleus(m.features, "Columnar")
    result = LibPQ.execute(m.conn, "SELECT COLUMNAR_AVG(\$1, \$2)", [table, column])
    v = _float(result)
    return v === nothing ? 0.0 : v
end

"""COLUMNAR_MIN(table, column) → Any"""
function columnar_min(m::ColumnarModel, table::String, column::String)
    require_nucleus(m.features, "Columnar")
    result = LibPQ.execute(m.conn, "SELECT COLUMNAR_MIN(\$1, \$2)", [table, column])
    val = first(result)[1]
    return ismissing(val) ? nothing : val
end

"""COLUMNAR_MAX(table, column) → Any"""
function columnar_max(m::ColumnarModel, table::String, column::String)
    require_nucleus(m.features, "Columnar")
    result = LibPQ.execute(m.conn, "SELECT COLUMNAR_MAX(\$1, \$2)", [table, column])
    val = first(result)[1]
    return ismissing(val) ? nothing : val
end

File: python/neutron/nucleus/columnar.py
"""Columnar model — wraps Nucleus COLUMNAR_* SQL functions."""

from __future__ import annotations

from typing import Any, cast

from neutron.nucleus._exec import Executor, require_nucleus
from neutron.nucleus.client import Features


class ColumnarModel:
    """Columnar analytics operations over Nucleus.

    Usage::

        await db.columnar.insert("metrics", {"ts": 1700000000, "value": 42.5})
        total = await db.columnar.count("metrics")
        avg = await db.columnar.avg("metrics", "value")
    """

    def __init__(self, executor: Executor, features: Features) -> None:
        self._exec = executor
        self._features = features

    def _require(self) -> None:
        require_nucleus(self._features, "Columnar")

    async def insert(self, table: str, values: dict[str, Any]) -> bool:
        """Insert a row into a columnar table.

        COLUMNAR_INSERT is variadic: (table, col1, val1, col2, val2, ...)
        and returns the text 'OK' on success.
        """
        self._require()
        args: list[Any] = [table]
        for column, value in values.items():
            args.append(column)
            args.append(value)
        placeholders = ", ".join(f"${i + 1}" for i in range(len(args)))
        result = await self._exec.fetchval(
            f"SELECT COLUMNAR_INSERT({placeholders})", *args
        )
        return result == "OK"

    async def count(self, table: str) -> int:
        """Return the row count of a columnar table."""
        self._require()
        return cast("int", await self._exec.fetchval("SELECT COLUMNAR_COUNT($1)", table))

    async def sum(self, table: str, column: str) -> float:
        """Return the sum of a column."""
        self._require()
        return cast(
            "float",
            await self._exec.fetchval(
                "SELECT COLUMNAR_SUM($1, $2)", table, column
            )
        )

    async def avg(self, table: str, column: str) -> float | None:
        """Return the average of a column."""
        self._require()
        return cast(
            "float | None",
            await self._exec.fetchval(
                "SELECT COLUMNAR_AVG($1, $2)", table, column
            )
        )

    async def min(self, table: str, column: str) -> Any:
        """Return the minimum value of a column."""
        self._require()
        return await self._exec.fetchval(
            "SELECT COLUMNAR_MIN($1, $2)", table, column
        )

    async def max(self, table: str, column: str) -> Any:
        """Return the maximum value of a column."""
        self._require()
        return await self._exec.fetchval(
            "SELECT COLUMNAR_MAX($1, $2)", table, column
        )
